lib/AST/Decl.cpp

```diff
@@ -2426,6 +2448,124 @@ ValueDecl::getFormalAccessScope(const DeclContext *useDC,
   llvm_unreachable("unknown access level");
 }
 
+AccessScope
+ValueDecl::getFormalAccessScope(const DeclContext *useDC,
+                                bool treatUsableFromInlineAsPublic) const {
+  return getAccessScopeForFormalAccess(this, getFormalAccess(), useDC,
+                                       treatUsableFromInlineAsPublic);
+}
+
+/// Checks if \p VD may be used from \p useDC, taking \@testable imports into
+/// account.
+///
+/// Whenever the enclosing context of \p VD is usable from \p useDC, this
+/// should compute the same result as checkAccess, below, but more slowly.
+///
+/// See ValueDecl::isAccessibleFrom for a description of \p forConformance.
+static bool checkAccessUsingAccessScopes(const DeclContext *useDC,
+                                         const ValueDecl *VD,
+                                         AccessLevel access,
+                                         bool forConformance) {
+  bool treatUsableFromInlineAsPublic = false;
+
+  // FIXME: Swift 4.1 allowed accessing protocol extension methods that were
+  // marked 'public' if the protocol was '@_versioned' (now
+  // '@usableFromInline'). Which works at the ABI level, so let's keep
+  // supporting that here by passing 'treatUsableFromInlineAsPublic'.
+  if (!forConformance &&
+      VD->getDeclContext()->getAsProtocolOrProtocolExtensionContext()) {
+    treatUsableFromInlineAsPublic = true;
+  }
+
+  AccessScope accessScope =
+        getAccessScopeForFormalAccess(VD, access, useDC,
+                                      treatUsableFromInlineAsPublic);
+  return accessScope.getDeclContext() == useDC ||
+         AccessScope(useDC).isChildOf(accessScope);
+}
+
+/// Checks if \p VD may be used from \p useDC, taking \@testable imports into
+/// account.
+///
+/// When \p access is the same as `VD->getFormalAccess()` and the enclosing
+/// context of \p VD is usable from \p useDC, this ought to be the same as
+/// getting the AccessScope for `VD` and checking if \p useDC is within it.
+/// However, there's a source compatibility hack around protocol extensions
+/// that makes it not quite the same.
+///
+/// See ValueDecl::isAccessibleFrom for a description of \p forConformance.
+static bool checkAccess(const DeclContext *useDC, const ValueDecl *VD,
+                        AccessLevel access, bool forConformance) {
+  if (!useDC)
+    return access >= AccessLevel::Public;
+
+  auto *sourceDC = VD->getDeclContext();
+
+  if (!forConformance && sourceDC->getAsProtocolOrProtocolExtensionContext()) {
+    // Skip the fast path below and just compare access scopes.
+    return checkAccessUsingAccessScopes(useDC, VD, access, forConformance);
+  }
+
+  switch (access) {
```

Comment:
It's at the call site, but it's probably worth including here too.

Access scope computation always walks the full DeclContext chain unless something is local, because it's recursive. It is something we could consider caching, though.